### ISSUE TITLE
Consder allowing self-referencing `insert`

### ISSUE THREAD

**Issue by elprans** [edgedb/edgedb#3960]
Currently, self-referencing inserts are prohibited, i.e. the following will throw:

```
insert User { modified_by := User }
```

This is a useful pattern sometimes, and a more elaborate example might include a wider reference loop.  It's possible to work around this with an immediate `update`, e.g:

```
with new_user := (insert User) update new_user set { modified_by := new_user }
```

though this will only work if `modified_by` is not `required`.

Thus, the most ergonomic way of allowing this seems to be to permit self-referencing inserts.

The implementation strategy would obviously be to effectively simulate the `update` workaround and handle `required` links by providing a temporary placeholder identifier.

**Comment by nsidnev** [comment]
I tried to take a look into it, but it looks like the workaround itself also has an issue (though it probably has no relation to this one, and may require a separate bug):

```
edgedb> start migration to { module default { type User { link modified_by -> User; } } };
OK: START MIGRATION
edgedb[tx]> populate migration;
OK: POPULATE MIGRATION
edgedb[tx]> commit migration;
OK: COMMIT MIGRATION
edgedb> with new_user := (insert User) update new_user set { modified_by := new_user };
{}
edgedb> select User { modified_by };
{default::User {modified_by: {}}}
edgedb> with new_user := (insert User) update new_user set { modified_by := (select User filter .id = new_user.id) };
{}
edgedb> select User { modified_by };
{default::User {modified_by: {}}, default::User {modified_by: {}}}
```

Not sure if I'm doing something wrong. Tried with [latest master](https://github.com/edgedb/edgedb/commit/c3a1e1c5c8d343dd3d02d5233cb13a9798770262):
```
$ edb server --version
edb, version 2.0-dev.6664+d2022062503.gc3a1e1c5c.cv202206221500.r202206261125.tpa4dmxzwgqwxk3tlnzxxo3rnnruw45lyfvtw45i.blocal.sb4a2d87
```

**Comment by nsidnev** [comment]
@elprans I think this feature, as well as the workaround, cannot be implemented at this time. I tried to run a similar query with SQL and it doesn't work either.

Table:
```sql
create table users (id UUID primary key, modified_by UUID, foreign key (modified_by) references users(id)); 
```

Query:
```sql
with new_user as (insert into users(id) values (uuid_generate_v4()) returning id)
update users as u
set modified_by = new_user.id
from new_user
where u.id = new_user.id;
```

The result is a new user record, but the `modified_by` field has `NULL` value.

I discussed this with @msullivan and he agreed with my thoughts. At this point I suggest adding a test with `xfail` for this case until it can be implemented, but maybe you have other thoughts that might help here?

**Comment by elprans** [comment]
Yeah, it appears you cannot update the just-inserted tuple.  Bummer.

**Comment by CodesInChaos** [comment]
Couldn't you first generate the id, and then insert the object setting both the object's id and the link target to that id at the same time? Since the id comes from a trusted source, checking if there are already objects for it should not be necessary.
